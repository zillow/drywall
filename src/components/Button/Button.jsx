import PropTypes from 'prop-types';
import styled from 'styled-components';

export const BUTTON_TYPES = {
    primary: 'primary',
    secondary: 'secondary',
    caution: 'caution',
};

export const BUTTON_SIZES = {
    sm: 'sm',
    md: 'md',
    lg: 'lg',
};

const Button = styled.button`
    ${props => props.theme && props.theme.ns && props.theme.ns().Button};
`;

Button.propTypes = {
    /** A fluid button will take up all horizontal space */
    fluid: PropTypes.bool,
    /** Inverse colors for use on dark or colored backgrounds */
    inverse: PropTypes.bool,
    /** The size of the button */
    size: PropTypes.oneOf(Object.keys(BUTTON_SIZES)),
    /** The type of the button */
    buttonType: PropTypes.oneOf(Object.keys(BUTTON_TYPES)),
    /**
     * Use the styled-components
     * [polymorphic "as" prop](https://www.styled-components.com/docs/api#as-polymorphic-prop)
     * to change the HTML tag of the component.
     *
     * This can be used to apply `Button` styling to HTML anchors.
     */
    as: PropTypes.string,
};

Button.defaultProps = {
    fluid: false,
    inverse: false,
    size: BUTTON_SIZES.md,
    buttonType: BUTTON_TYPES.secondary,
    as: 'button',
};

/** @component */
export default Button;
export const StyledButton = Button;
