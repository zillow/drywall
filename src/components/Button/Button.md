#### Button types
```jsx
<Button buttonType="primary">Primary label</Button>{' '}
<Button buttonType="secondary">Secondary button</Button>{' '}
<Button buttonType="caution">Caution button</Button>{' '}
```

#### Disabled buttons
```jsx
<Button buttonType="primary" disabled>Primary label</Button>{' '}
<Button buttonType="secondary" disabled>Secondary button</Button>{' '}
<Button buttonType="caution" disabled>Caution button</Button>{' '}
```

#### Fluid button
```jsx
<Button fluid>Fluid button</Button>
```

#### Button sizes
```jsx
<Button size="sm">Small button</Button>{' '}
<Button size="md">Medium button</Button>{' '}
<Button size="lg">Large button</Button>
```

#### Inversed buttons for colored backgrounds or background images

```jsx
import TransparencyWrapper from '../../styleguidist/TransparencyWrapper';

<TransparencyWrapper dark style={{ padding: '10px' }}>
    <Button buttonType="primary" inverse>Primary label</Button>{' '}
    <Button buttonType="secondary" inverse>Secondary button</Button>
</TransparencyWrapper>
```

#### Anchors as Buttons

```jsx
<Button as="a" href="https://www.zillow.com/" buttonType="primary">Zillow anchor</Button>{' '}
<Button as="a" href="https://www.zillow.com/" buttonType="secondary">Zillow anchor</Button>{' '}
<Button as="a" href="https://www.zillow.com/" buttonType="caution">Zillow anchor</Button>
```
