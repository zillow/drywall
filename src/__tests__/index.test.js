import React from 'react';
import renderer from 'react-test-renderer';
import Button, { StyledButton } from '../components/Button/Button';
import Bootstrap from '../themes';

describe('Themed Styled Button', () => {
    it('displays a button with requested text', () => {
        const component = renderer.create(<StyledButton theme={Bootstrap}>Click Me!</StyledButton>);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('displays a secondary button with requested text', () => {
        const component = renderer.create(
            <StyledButton theme={Bootstrap} type="secondary">
                ⊙_ʘ
            </StyledButton>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});

describe('Themed Button', () => {
    it('displays a button with requested text', () => {
        const component = renderer.create(<Button theme={Bootstrap}>Click Me!</Button>);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('displays a secondary button with requested text', () => {
        const component = renderer.create(
            <Button theme={Bootstrap} type="secondary">
                ⊙_ʘ
            </Button>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});
