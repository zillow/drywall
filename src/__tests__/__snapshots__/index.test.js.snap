// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Themed Button displays a button with requested text 1`] = `
.c0 {
  display: inline-block;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  -webkit-transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  color: #fff;
  background-color: #007bff;
  border-color: #007bff;
}

.c0:hover {
  color: #fff;
  background-color: #0069d9;
  border-color: #0062cc;
}

<button
  className="c0"
  type="primary"
>
  Click Me!
</button>
`;

exports[`Themed Button displays a secondary button with requested text 1`] = `
.c0 {
  display: inline-block;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  -webkit-transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  color: #fff;
  background-color: #6c747d;
  border-color: #6c757d;
}

.c0:hover {
  color: #fff;
  background-color: #5a6268;
  border-color: #545b62;
}

<button
  className="c0"
  type="secondary"
>
  ⊙_ʘ
</button>
`;

exports[`Themed Styled Button displays a button with requested text 1`] = `
.c0 {
  display: inline-block;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  -webkit-transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  color: #fff;
  background-color: #6c747d;
  border-color: #6c757d;
}

.c0:hover {
  color: #fff;
  background-color: #5a6268;
  border-color: #545b62;
}

<button
  className="c0"
>
  Click Me!
</button>
`;

exports[`Themed Styled Button displays a secondary button with requested text 1`] = `
.c0 {
  display: inline-block;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5;
  border-radius: 0.25rem;
  -webkit-transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  transition: color .15s ease-in-out,background-color .15s ease-in-out,border-color .15s ease-in-out,box-shadow .15s ease-in-out;
  color: #fff;
  background-color: #6c747d;
  border-color: #6c757d;
}

.c0:hover {
  color: #fff;
  background-color: #5a6268;
  border-color: #545b62;
}

<button
  className="c0"
  type="secondary"
>
  ⊙_ʘ
</button>
`;
